php/libraries/User.class.inc

```diff
@@ -411,89 +422,76 @@ class User extends UserPermissions
-        $CharTypes += preg_match('/[0-9]+/', $password);
-        $CharTypes += preg_match('/[A-Za-z]+/', $password);
-        $CharTypes += preg_match('/[!\\\$\^@#%&\*\(\)]+/', $password);
-        if ($CharTypes < 3) {
-            return false;
-        }
-
-        // compare password to values
-        foreach ($values as $key => $value) {
-            switch ($operators[$key]) {
-            case '==':
-                // They must be equal, so the whole thing is false
-                // if they're not.
-                if ($password != $value) {
-                    return false;
-                }
-                break;
-            case '!=':
-                if ($password == $value) {
-                    return false;
-                }
-                break;
-            default:
-                throw new \LorisException("Invalid operator in isPasswordStrong");
-            }
-        }
-
-        return true;
+    function checkPasswordStrength(string $password): array
+    {
+        $user = \User::singleton();
+        $data = $user->getData();
+        /* This is an array of information related to the user. This is used by
+         * Zxcvbn to check if a user is doing something foolish like making
+         * their password 'johnnyloris'.
+         * We should add words as we think of them.
+         */
+        $userInputs = array(
+                       $user->getUsername(),
+                       $data['Email'],
+                       $data['Real_name'],
+                       'loris',
+                       'brain',
+                       'mcgill',
+                      );
+
+        $zxcvbn = new Zxcvbn();
+        return $zxcvbn->passwordStrength($password, $userInputs);
     }
 
+
     /**
-     * Check if user belongs to DCC
+     * Validate a potential password to check whether it meets our requirements,
+     * including length as well as strength measured by Zxcvbn. Also makes sure
+     * that the new password is different from the old one.
      *
-     * @return boolean
+     * @param string $newPassword The password to validate.
+     *
+     * @return array A bool and string array representing whether a password is
+     *                          valid or not and why we have marked it as such.
      */
-    function isUserDCC()
+    function validatePassword(string $newPassword): array
     {
-        //DCC site = 1 by LORIS convention
-        return ($this->userInfo['CenterID']== 1);
+        $result        = array(
+                          'valid' => false,
+                          'why'   => '',
+                         );
+        $minimumLength = 8;
+        if (strlen($newPassword) < $minimumLength) {
+            $result['why'] = 'Passphrase is too short.';
+            return $result;
+        }
+        $data = \User::singleton()->getData();
+        if (password_verify($newPassword, $data['Password_hash'])) {
+            $result['why'] = 'You cannot keep the same passphrase.';
```

Comment:
should be "... same password."

Reply:
No, there is no standard right now. I like passphrase but don't feel that strongly about it. I'll bring it up in the LORIS meeting.

Reply:
Then, just to keep the same word everywhere, I would use "password"

Reply:
Would also recommend "password" for consistency.